### PR TITLE
1.25

### PR DESCRIPTION
**Pull request by errm**
- Upgrades to k8s 1.25
- Moves to a much more minimal setup
  - ASG node groups are no longer supported
  - Karpenter must be used to provision nodes
    - Karpenter is run via fargate
  - All addons / cluster configuration is removed from this module.
    - Recommended to manage via flux or other gitops
  
#  📝 TODO

- Upgrade procedure / instructions!

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -49,60 +48,74 @@ resource "aws_cloudwatch_log_group" "control_plane" {
...
-  config  = local.config
-  replace = true
-  manifest = templatefile(
-    "${path.module}/storage_classes.yaml.tmpl",
-    {
-      provisioner = "ebs.csi.aws.com",
-      fstype      = "csi.storage.k8s.io/fstype: ${var.pv_fstype}",
-    }
-  )
+resource "kubernetes_config_map_v1_data" "aws_auth" {
```
Comment (jportasa) [inline main.tf:98]: I see this block is very similar to `resource "kubernetes_config_map" "aws_auth"` why we have this twice?

File: main.tf [hunk 2]
```diff
@@ -49,60 +48,74 @@ resource "aws_cloudwatch_log_group" "control_plane" {
...
+          rolearn  = aws_iam_role.karpenter_fargate.arn
+          username = "system:node:{{SessionName}}"
+          groups = [
+            "system:bootstrappers",
+            "system:nodes",
+            "system:node-proxier",
+          ]
+        },
+        {
+          rolearn  = aws_iam_role.karpenter_node.arn
```
Comment (jportasa) [inline main.tf:69]: In web-platform-X  we use` instanceProfile: EKSNode` [here](https://github.com/cookpad/global-web-platform/blob/main/k8s/infrastructure/base/karpenter/provisioner-default.yaml#L35) we will have to change it in the provisioner?

File: karpenter_interruption_queue.tf [hunk 3]
```diff
@@ -0,0 +1,74 @@
+resource "aws_sqs_queue" "karpenter_interruption" {
+  name                      = "Karpenter-${var.name}"
```
Comment (jportasa) [inline karpenter_interruption_queue.tf:2]: maybe more clear karpepenter-spot-node-interruption-${var.name}?

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,86 @@
+variable "name" {
```
Comment (jportasa) [inline variables.tf:1]: add a description

File: examples/cluster/main.tf [hunk 5]
```diff
@@ -1,35 +1,32 @@
-terraform {
```
Comment (jportasa) [inline examples/cluster/main.tf:1]: create something like this https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/versions.tf

File: versions.tf [hunk 6]
```diff
@@ -1,3 +1,15 @@
 terraform {
   required_version = ">= 1.0"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">= 4.47.0"
```
Comment (jportasa) [inline versions.tf:7]: 5.1.0 maybe?

File: versions.tf [hunk 7]
```diff
@@ -1,3 +1,15 @@
...
+    }
+
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = ">= 2.10"
```
Comment (jportasa) [inline versions.tf:12]: 2.21.1 is the latest

File: karpenter_fargate.tf [hunk 8]
```diff
@@ -0,0 +1,39 @@
+resource "aws_eks_fargate_profile" "karpenter" {
+  cluster_name           = aws_eks_cluster.control_plane.name
+  fargate_profile_name   = "Karpenter-${var.name}"
+  pod_execution_role_arn = aws_iam_role.karpenter_fargate.arn
+  subnet_ids             = values(var.vpc_config.private_subnet_ids)
+
+  selector {
+    namespace = "karpenter"
+    labels    = {}
```
Comment (jportasa) [inline karpenter_fargate.tf:9]: we want as well flux-system namespace to run on Fargate

File: versions.tf (hunk 7 above)
Comment (errm) [reply versions.tf:12]: Same

File: main.tf (hunk 2 above)
Comment (errm) [reply main.tf:69]: That is the intention, but we can add the old profile name to the map during transition.

File: karpenter_interruption_queue.tf [hunk 9]
```diff
@@ -0,0 +1,74 @@
+resource "aws_sqs_queue" "karpenter_interruption" {
```
[the remark this answers is not in the document]
Comment (errm) [reply karpenter_interruption_queue.tf:1]: I don't think it makes it clearer, this queue will have a number of different events - not only spot interruptions.

**Pull request merged**